### PR TITLE
[GO] TLS for REST and gRPC

### LINKED ISSUE #1461: [Go] REST TLS
## 📖 Description

Implement TLS within the REST transport. Will need options added to the `config.Options` struct. Should have `rest` prefixed in all the env names so it can coexist with the grpc options.

---

## 🎯 Goal / Outcome

REST TLS Support

---

## ✅ Definition of Done (DoD)
- [ ] REST TLS Connection
- [ ] Edge cases are handled
- [ ] Tests are added or updated (if applicable)
- [ ] Documentation is updated (if needed)
- [ ] Code is reviewed and approved -->

---

## 🛠️ Implementation Notes (Optional)

---

## 🔗 Related Issues / Links

#1462 
---
## 🚧 Risks / Open Questions (Optional)

---

### LINKED ISSUE #1462: [Go] GRPC TLS
## 📖 Description

Implement TLS within the GRPC transport. Will need options added to the `config.Options` struct. Should have `grpc` prefixed in all the env names so it can coexist with the rest options.

---

## 🎯 Goal / Outcome

GRPC TLS Support

---

## ✅ Definition of Done (DoD)
- [ ] GRPC TLS Connection
- [ ] Edge cases are handled
- [ ] Tests are added or updated (if applicable)
- [ ] Documentation is updated (if needed)
- [ ] Code is reviewed and approved

---

## 🛠️ Implementation Notes (Optional)

---

## 🔗 Related Issues / Links

#1462 

---
## 🚧 Risks / Open Questions (Optional)

---

### PR DESCRIPTION
**Pull request by nelgd**
## 📖 Description

Adds optional TLS and mutual TLS (mTLS) support to both the REST and gRPC transports in the Go SDK. A shared `config.TLSOptions` struct is embedded in `RestOptions` and `GrpcOptions`, configurable via transport-prefixed environment variables and CLI Flags

Env Variable: `CATENA_REST_TLS_*`, `CATENA_GRPC_TLS_*`

CLI Flags: `--rest-tls-*`, `--grpc-tls-*`

---

## 🎯 Goal / Outcome

- Servers built with the Go SDK can serve HTTPS on the REST transport and TLS-encrypted gRPC, including optional client-certificate verification (mTLS), configured entirely through the existing config system.

---

## ✅ Definition of Done (DoD)

- [x] REST transport serves TLS via `http.Server.ServeTLS` when `REST_TLS_ENABLED` is set
- [x] gRPC transport serves TLS via `grpc.Creds` when `GRPC_TLS_ENABLED` is set
- [x] mTLS supported on both transports (`*_TLS_MUTUAL_AUTH` + `*_TLS_CA_FILE`, enforced with `RequireAndVerifyClientCert`)
- [x] Tests added: config loading (env + flags), `ServerTLSConfig()` validation cases, `LogValue` output, and TLS/mTLS integration tests for both transports (handshake success, plaintext/insecure rejection, missing client cert rejection, bad-config startup failure)
- [x] Documentation updated: `config/README.md` env/flag tables, `transports/README.md`, `REST.md`, `GRPC.md`
---

## 🛠️ Implementation Notes

- **Shared `TLSOptions` struct** (`pkg/config/config.go`) embedded in both transport option structs; `ServerTLSConfig()` (`pkg/config/tls.go`) centralizes validation and builds a `*tls.Config` with `MinVersion: TLS 1.2`. Returns `(nil, nil)` when disabled so callers keep the plaintext path.
- **REST**: TLS config is built at `Start` so misconfiguration returns a synchronous startup error.
- **gRPC**: `grpc.Creds` must be supplied at server construction, but `NewTransport` cannot return an error, so a TLS config failure is recorded in `initErr` and surfaced from `Start`. `Shutdown` is nil-safe for this case.
- **Test certificates**: new internal package `pkg/internal/testcerts` generates ephemeral self-signed CA/server/client certs at test runtime — no fixture files or long-lived keys checked in.
- **`TLSOptions` re-exported** from the top-level `catena` package for SDK consumers.
- **CI toolchain fix** (`.devcontainer/go/toolchain.env`): `IMAGE_VERSION` v0.0.1 → v0.0.2 forces the toolchain-delta workflow to rebuild the Go image with Go 1.26.0; `OPENSSL_VERSION` updated because Ubuntu removed `3.0.13-0ubuntu3.11` from the archive on 2026-07-31, which broke the image build.

---

## 🔗 Related Issues / Links

Closes issues:
- #1461 
- #1462

### CODE REVIEW

**Comment by nelgd** [comment]
Had issues getting the GitHub build to pass from unit tests failing due to mismatched go versions. I made changes to the tool chain for the dev container and that seemed to fix it. Not sure what's up with that.

```
IMAGE_VERSION="v0.0.1" --> IMAGE_VERSION="v0.0.2"
OPENSSL_VERSION=3.0.13-0ubuntu3.11 --> OPENSSL_VERSION=3.0.13-0ubuntu3.12
```

File: sdks/go/pkg/config/README.md [hunk 1]
```diff
@@ -253,6 +272,25 @@ All options can be configured via env and CLI.
 
 If `--log-level` is explicitly set, it takes priority over these shortcuts.
 
+## Transport TLS
+
+The REST and gRPC transports each accept an independent `TLSOptions` (`Rest.TLS` / `Grpc.TLS`), so one can run with TLS while the other runs plaintext. All file options take full paths to PEM-encoded files.
+
+- `*_TLS_ENABLED`: turn TLS on for that transport's listener (default off).
+- `*_TLS_CERT_FILE` / `*_TLS_KEY_FILE`: server certificate and private key. Both are required when TLS is enabled; the transport fails to start if either is missing or unreadable.
+- `*_TLS_MUTUAL_AUTH`: require clients to present a certificate signed by the CA in `*_TLS_CA_FILE` (mTLS).
```
Comment (abrown-RV) [inline sdks/go/pkg/config/README.md]: (default off)

File: sdks/go/pkg/internal/testcerts/testcerts.go [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+	CACertFile string
+	// ServerCertFile / ServerKeyFile are a server key pair valid for
+	// "localhost" and 127.0.0.1/::1.
+	ServerCertFile string
+	ServerKeyFile  string
+	// ClientCertFile / ClientKeyFile are a client key pair for mTLS tests.
+	ClientCertFile string
+	ClientKeyFile  string
+}
+
```
Comment (abrown-RV) [inline sdks/go/pkg/internal/testcerts/testcerts.go:73]: add this to the coverage exclusion in the Makefile, its testing utils, no need to cover

File: sdks/go/pkg/config/tls.go [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+		if err != nil {
+			return nil, fmt.Errorf("failed to read client CA file %q: %w", o.CAFile, err)
+		}
+		caPool := x509.NewCertPool()
+		if !caPool.AppendCertsFromPEM(caPEM) {
+			return nil, fmt.Errorf("client CA file %q contains no valid PEM certificates", o.CAFile)
+		}
+		tlsConfig.ClientCAs = caPool
+		tlsConfig.ClientAuth = tls.RequireAndVerifyClientCert
+	} else if o.CAFile != "" {
```
Comment (abrown-RV) [inline sdks/go/pkg/config/tls.go]: drop this if branch, we're not suppose to log like this. This can just be a silent skip if MutualAuth is off but the CAFile is set. Put something documenting the field is ignored if MutualAuth is false.

File: sdks/go/pkg/config/config.go [hunk 4]
```diff
@@ -57,10 +57,35 @@ type RuntimeOptions struct {
...
+	Enabled bool
+	// CertFile is the path to the PEM-encoded server certificate.
+	// Required when Enabled is true.
+	CertFile string
+	// KeyFile is the path to the PEM-encoded server private key.
+	// Required when Enabled is true.
+	KeyFile string
+	// CAFile is the path to a PEM-encoded CA bundle used to verify client
+	// certificates. Required when MutualAuth is true; ignored otherwise.
+	CAFile string
```
Comment (abrown-RV) [inline sdks/go/pkg/config/config.go]: `CAFile` is a little ambiguous, make sure to update env/cli as well
```suggestion
	ClientCAFile string
```

**Review by abrown-RV — CHANGES_REQUESTED** [review]
Nice work Nelson. Just some minor tweaks.

**Comment by nelgd** [comment]
Not sure why the checks are failing. I don't think its an issue with my code

**Review by abrown-RV — APPROVED** [review]
